src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -792,10 +802,15 @@ static int client_message_init(
         packet = malloc0(size);
         if (!packet)
                 return -ENOMEM;
-
-        r = dhcp_message_init(&packet->dhcp, BOOTREQUEST, client->xid, type,
-                              client->arp_type, client->hw_addr.length, client->hw_addr.bytes,
-                              optlen, &optoffset);
+        if (client->bootp) {
+                r = bootp_message_init(&packet->dhcp, BOOTREQUEST, client->xid, client->arp_type,
+                                       client->hw_addr.length, client->hw_addr.bytes);
+                optoffset = 0;
```

Comment:
maybe switch these two lines. usually we want to keep function calls and their immediate error handling together. i.e. the bootp_message_init() is done here, but the `if (r < 0)` is further down. here you insert another line in between.